### ISSUE TITLE
Wrapper helpers using reflection

### ISSUE THREAD

**Issue by ktsivkov** [newrelic/go-agent#970]
Provide helper functions to enable easier and cleaner way to start transactions and segments.

## Summary

By my experience most of the time the applications that require transactions they are quite straight forward.
- Starting a transaction
- Starting Segments
- Ending segments
- Ending transactions

## Desired Behaviour

Optimally will be nice to have a way to wrap application specific functions that when invoked will start transactions and segments without coupling the application code to the agent's behavior, and without changing their underlying type.

To achieve the wrapping behavior we can use reflections, and to avoid changing the underlying type of the "wrapees" we can use generics.

## Possible Solution

I can think of two solutions, the easiest one could be the addition of helper functions WrapTransactionFunc and WrapSegmentFunc, and they can be implemented like this:
```go
func WrapTransactionFunc[V any](app *Application, wrapee V, transactionName string, traceOptions ...TraceOption) V {
	typ := reflect.TypeOf(wrapee)
	if typ.Kind() != reflect.Func {
		panic("wrapee must be a func")
	}

	wrappedFunc := func(args []reflect.Value) []reflect.Value {
		tx := app.StartTransaction(transactionName, traceOptions...)
		defer tx.End()

		return reflect.ValueOf(wrapee).Call(args)
	}

	fn := reflect.MakeFunc(typ, wrappedFunc)
	return fn.Interface().(V)
}


func WrapSegmentFunc[V any](tx *Transaction, wrapee V, segmentName string) V {
	typ := reflect.TypeOf(wrapee)
	if typ.Kind() != reflect.Func {
		panic("wrapee must be a func")
	}

	wrappedFunc := func(args []reflect.Value) []reflect.Value {
		sg := tx.StartSegment(segmentName)
		defer sg.End()

		return reflect.ValueOf(wrapee).Call(args)
	}

	fn := reflect.MakeFunc(typ, wrappedFunc)
	return fn.Interface().(V)
}
```

**Comment by iamemilio** [comment]
This would rack up quite a bit of overhead. Have you played with github.com/newrelic/go-easy-instrumentation at all?

**Comment by ktsivkov** [comment]
@iamemilio No, but just checked the readme file. It looks quite interesting, I will definitely have a look!

**Comment by iamemilio** [comment]
Play with it and don't hesitate to leave us some feedback

**Comment by iamemilio** [comment]
I am going to go ahead and close this if thats ok. The issue with using reflection is that its fairly expensive. We are already adding a bunch of extra processing by adding the go agent to an application. Our job is to really keep that to a minimum. We encourage users to create solutions that work for them on top of the tools we provide.
